pkg/agent/stream.go

```diff
@@ -111,6 +101,42 @@ func NewRedialableAgentClient(address, agentID string, cs *ClientSet, opts ...gr
 	return c, nil
 }
 
+func (c *RedialableAgentClient) copyRedialableAgentClient() *RedialableAgentClient {
+	out := &RedialableAgentClient{
+		cs:                      c.cs,
```

Comment:
Copying the object copies fields we don't want to copy. That then triggers the race warning on fields we didn't even want to copy. So its manually copy what we actually want and not copying the things we don't want. Or we start layering the copy in locks for things we don't care about.

Reply:
Got it. It's good that we don't copy the locks, otherwise this would be recursive.

It seems that the reconnecting code is causing more and more maintenance burden. Maybe we should remove the functionality entirely.

The advantage of the reconnecting is that it will retry the current packet so it surfaces less errors to the caller. Maybe now the maintenance burden finally overgrows the advantage. What do you think? I know @Jefftree and @dberkov advocates to remove this code as well.

Reply:
I can do that in a followup PR if we all agree to remove the code.

Reply:
I'm fine with removing that if we're finding it too much of a burden. It also occurred to me that the old code is actually more dangerous. It was relying on a shallow copy. If we add more fields (slices, pointers etc) and don't fix this method, we end up with some fairly nasty bugs. The new and old objects look fine but have shared state. At least this code breaks obviously if you add a field and don't fix this method. The field just isn't set.

Reply:
I agree that if we want to remove that functionality it should be a separate PR. Honestly I would prefer to break this PR up into pieces. Unfortunately we wouldn't be able to turn on race testing until all the pieces merged.

Reply:
I'm generally ok with it. Retrying packets seems like a nice to have. I would say as long as we have a metric to measure how often that case is hit I'm fine with removing it. That way if it goes from nice to have to a something we need to fix, we will have a way to know.

Reply:
After looking at all the fields copied, it looks like they can all be obtained from the clientset with the exception of serverCount. This shouldn't block the PR but perhaps we don't actually need a field by field copy and can just call `NewAgentClient` from the clientset.